File: AstralSDK/Sources/AstralSDK/Core/AstralError.swift
import Foundation

/// All recoverable errors in AstralSDK.
public enum AstralError: Error, Sendable, CustomStringConvertible {
    case notInitialized
    case keyGenerationFailed
    case signingFailed
    case verificationFailed
    case noTransportAvailable
    case bleUnavailable
    case relayUnavailable
    case packetEncodingFailed
    case packetDecodingFailed
    case invalidQRCode
    case paymentFailed(String)
    case timeout
    case duplicate

    public var description: String {
        switch self {
        case .notInitialized:       return "SDK not initialized — call start() first"
        case .keyGenerationFailed:  return "Failed to generate keypair"
        case .signingFailed:        return "Transaction signing failed"
        case .verificationFailed:   return "Signature verification failed"
        case .noTransportAvailable: return "No transport available (BLE/Relay/Queue)"
        case .bleUnavailable:       return "Bluetooth is not available"
        case .relayUnavailable:     return "Relay server is unreachable"
        case .packetEncodingFailed: return "Failed to encode packet"
        case .packetDecodingFailed: return "Failed to decode packet"
        case .invalidQRCode:        return "Invalid or unrecognized QR code"
        case .paymentFailed(let m): return "Payment failed: \(m)"
        case .timeout:              return "Operation timed out"
        case .duplicate:            return "Duplicate transaction"
        }
    }
}

File: android/astralsdk/src/main/java/com/astralnetwork/sdk/core/AstralError.kt
package com.astralnetwork.sdk.core

/**
 * AstralError — SDK error types.
 * Mirrors iOS AstralError enum for cross-platform consistency.
 */
sealed class AstralError(message: String) : Exception(message) {
    object NotInitialized : AstralError("SDK not initialized")
    object BleUnavailable : AstralError("Bluetooth unavailable")
    object RelayUnavailable : AstralError("Relay server unavailable")
    object SigningFailed : AstralError("Transaction signing failed")
    object VerificationFailed : AstralError("Signature verification failed")
    object Duplicate : AstralError("Double-spend: transaction already seen")
    object PacketDecodingFailed : AstralError("Packet decoding failed")
    class PaymentFailed(reason: String) : AstralError("Payment failed: $reason")
}
